### PR TITLE
Improvement to High Saliency Pointing Game

### PR DESCRIPTION
**Pull request by natalie-isak**
Aligning values of ground truth bounding box with the correct x and y channels. 

When iterating through a 2D array structure, the first loop will iterate through the y axis. The second loop will iterate through the x axis. Therefore, to correctly index to the ground truth bounding box, the following logic is needed:

`gt_bbox[gt_bbox[1]:gt_bbox[3]+1, gt_bbox[0]:gt_bbox[2]+1]`

instead of 

`gt_bbox[gt_bbox[0]:gt_bbox[2]+1, gt_bbox[1]:gt_bbox[3]+1]`

### CODE REVIEW

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 1]
```diff
@@ -148,15 +154,15 @@ def calculate_gt_salient_pixel_overlap(self,
...
         gt_mask = torch.zeros_like(saliency_scores, dtype=torch.bool)
-        gt_mask[gt_bbox[0]:gt_bbox[2], gt_bbox[1]:gt_bbox[3]] = True
+        gt_mask[gt_bbox[1]:gt_bbox[3]+1, gt_bbox[0]:gt_bbox[2]+1] = True
 
         positive_mask = torch.gt(saliency_scores, 0)
         positive_gt_mask = torch.logical_and(positive_mask, gt_mask)
-        good = positive_gt_mask.sum()
-        total = positive_mask.sum()
+        good = torch.count_nonzero(positive_gt_mask)
+        total = torch.count_nonzero(positive_mask)
```
Comment (cjbarberan) [inline python/vision_explanation_methods/evaluation/pointing_game.py:166]: Wouldnt the gt_mask and not the positive mask? The denominator is the issue when getting the values.

Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:166]: Right now the return value is # of salient pixels in the bounding box / # of total salient pixels. 

postive_mask is essentially the same as gt_mask. 

Does that answer your question?

In the future, I can add additional metrics with different denomenators (ie divide by # of pixels in the gt bounding box, divide by the # of pixels in the segmented object, etc.)

Comment (cjbarberan) [reply python/vision_explanation_methods/evaluation/pointing_game.py:166]: This would not be the case. The denominator how you provided is the saliency scores thresholded at 0. This will result in the pointing game having a higher accuracy compared to the denominator of the gt_mask.

Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:166]: Right that makes sense, but in gt_mask there are artificially set "-1"s --> why would we want those to contribute to the sum?

Comment (cjbarberan) [reply python/vision_explanation_methods/evaluation/pointing_game.py:166]: The bounding box will not always be the best, but with having the denominator be the saliency score it will be skewed higher and the saliency may not reflect the object. In the bear example that I tried this, it gives a higher accuracy then what it is and in the bounding box provided it covers a good amount of the bear.

Comment (cjbarberan) [reply python/vision_explanation_methods/evaluation/pointing_game.py:166]: and with gt_mask created with torch.zeros_like, there should be no -1s?

Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py]: I think I understand what you are saying now. I was confusing gt_mask and saliency_scores. 

I updated the code to calculate the following: number of salient pixels in the gt bbox / number of pixels in the gt bbox

LMK if the code looks good to you.

Comment (cjbarberan) [reply python/vision_explanation_methods/evaluation/pointing_game.py]: Sounds good.

**Review by cjbarberan — APPROVED** [review]
Looks good.

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 2]
```diff
@@ -140,6 +146,9 @@ def calculate_gt_salient_pixel_overlap(self,
         """
         Calculate percent of overlap between salient pixels and gt bbox.
 
+        Formula: number of salient pixels in the gt bbox /
```
Comment (imatiach-msft) [inline python/vision_explanation_methods/evaluation/pointing_game.py:149]: nit: maybe specify gt = ground truth in the docs to make it clearer

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 3]
```diff
@@ -41,7 +42,7 @@ def pointing_game(self,
                       imagelocation: str,
                       index: int,
                       threshold: float = .8,
-                      num_masks: int = 100):
+                      num_masks: int = 300):
```
Comment (imatiach-msft) [inline python/vision_explanation_methods/evaluation/pointing_game.py:45]: nit: move numbers to constants, perhaps at top of file, eg:
NUM_MASKS = 300
and reuse here as:
num_masks: int = NUM_MASKS
and elsewhere

Comment (imatiach-msft) [reply python/vision_explanation_methods/evaluation/pointing_game.py:45]: then in other code that calls this method we can specify the constants where needed

File: tests/test_vision_explanation.py [hunk 4]
```diff
@@ -196,6 +197,43 @@ def test_vision_explain_loadmodel():
         os.remove(elt)
 
 
+def test_pointing_game():
```
Comment (imatiach-msft) [inline tests/test_vision_explanation.py:200]: should we parameterize this test for different thresholds/num masks or would that be too much?

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 5]
```diff
@@ -41,7 +42,7 @@ def pointing_game(self,
                       imagelocation: str,
                       index: int,
                       threshold: float = .8,
```
Comment (imatiach-msft) [inline python/vision_explanation_methods/evaluation/pointing_game.py:44]: nit: add some validation to throw ValueError if threshold is not between 0 and 1 and num_masks is a negative value

**Review by imatiach-msft — APPROVED** [review]
approved with some minor comments

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 2 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:149]: Added to class initializer!

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 6]
```diff
@@ -67,6 +72,12 @@ def pointing_game(self,
         :return: 2d matrix of highly salient pixels
         :rtype: List[Tensor]
         """
+        if not 0 <= threshold <= 1:
+            raise ValueError('Threshold parameter not a float \
+                             between 0 and 1.')
+        if not 0 < num_masks:
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:78]:
```suggestion
        if num_masks < 0:
```
Nit: simpler code?

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 7]
```diff
@@ -131,6 +142,11 @@ def visualize_highly_salient_pixels(self,
             plt_fig_axis=(fig, ax),
             use_pyplot=False
         )
+        x, y, width, height = 247, 192, 108, 301
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:145]: Why is this hard-coded here? Shouldn't this be inferred from the image or passed as input to the class?

**Review by Advitya17 — CHANGES_REQUESTED** [review]

File: tests/test_vision_explanation.py [hunk 8]
```diff
@@ -196,38 +200,142 @@ def test_vision_explain_loadmodel():
...
+                                          num_masks=num_masks)
+        overlap = pg.calculate_gt_salient_pixel_overlap(salient_scores,
+                                                        gt_bbox)
+
+        # calculated salient pixel overlap equals expected brute force
+        good = 0
+        total = 0
+        for iindex, i in enumerate(salient_scores[0]):
+            for jindex, j in enumerate(i):
+                if j > 0:
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:295]: Nit: can combine lines 295 & 296 into one if condition.

File: tests/test_vision_explanation.py [hunk 9]
```diff
@@ -196,38 +200,142 @@ def test_vision_explain_loadmodel():
...
+                    if (gt_bbox[1] <= iindex <= gt_bbox[3]
+                       and gt_bbox[0] <= jindex <= gt_bbox[2]):
+                        good += 1
+                if (gt_bbox[1] <= iindex <= gt_bbox[3]
+                   and gt_bbox[0] <= jindex <= gt_bbox[2]):
+                    total += 1
+        overlap_check = good/total
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:302]:
```suggestion
        overlap_check = good / total
```
Nit

File: tests/test_vision_explanation.py [hunk 10]
```diff
@@ -196,38 +200,142 @@ def test_vision_explain_loadmodel():
...
+        s = pg.pointing_game(imgpath, index)
+
+        # check that the saliency map exists and has 3 channels
+        assert (len(s) == 3)
+
+        # calculate overlap
+        v = pg.calculate_gt_salient_pixel_overlap(s, [244, 139, 428, 519])
+
+        # check that this is a percent value
+        assert (0 < v < 1)
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:341]:
```suggestion
        assert 0 < v < 1
```
Nit: don't need brackets

File: tests/test_vision_explanation.py [hunk 11]
```diff
@@ -196,38 +200,142 @@ def test_vision_explain_loadmodel():
...
+                                              number_of_classes=1000)
+
+        # find saliency scores for top 20% of salient pixels
+        # do this for the second object in an image
+        pg = PointingGame(detection_model)
+        index = 0
+        s = pg.pointing_game(imgpath, index)
+
+        # check that the saliency map exists and has 3 channels
+        assert (len(s) == 3)
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:335]:
```suggestion
        assert len(s) == 3
```
Nit: don't need brackets

**Review by Advitya17 — CHANGES_REQUESTED** [review]

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 7 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:145]: oops!! leftover code from testing, removed

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 12]
```diff
@@ -148,15 +172,15 @@ def calculate_gt_salient_pixel_overlap(self,
...
+        gt_mask[gt_bbox[1]:gt_bbox[3]+1, gt_bbox[0]:gt_bbox[2]+1] = True
 
         positive_mask = torch.gt(saliency_scores, 0)
         positive_gt_mask = torch.logical_and(positive_mask, gt_mask)
-        good = positive_gt_mask.sum()
-        total = positive_mask.sum()
+        good = torch.count_nonzero(positive_gt_mask)
+        total = torch.count_nonzero(gt_mask)
 
         return torch.div(good, total).item()
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:186]: Side note: Assuming `.item()` doesn't need device detachment with a GPU etc. Feel free to also look up online to generally verify.

**Review by Advitya17 — APPROVED** [review]

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 12 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:186]: good point - looks like since .item() changes the object type to no longer be a tensor, the change to CPU device is automatic. Reference - https://discuss.pytorch.org/t/does-item-automatically-move-the-data-to-the-cpu/69629
